[google_sign_in] Add google_sign_in package (#413)
flutter-tizen/plugins: packages/google_sign_in/README.md

```yaml
dependencies:
  google_sign_in: ^5.4.0
  google_sign_in_tizen: ^0.1.0
```

For detailed usage on how to use `google_sign_in`, see https://pub.dev/packages/google_sign_in#usage.

## Tizen integration

`google_sign_in_tizen` supports Google Sign-In for [TVs and Limited Input Devices](https://developers.google.com/identity/gsi/web/guides/devices), note that many API scopes are limited for these device types, see [supported scopes](https://developers.google.com/identity/protocols/oauth2/limited-input-device#allowedscopes).

The plugin uses [OAuth 2.0 Device Authorization Grant](https://datatracker.ietf.org/doc/html/rfc8628) (aka device flow) which has a different authorization flow than [Authorization Code Grant with PKCE](https://datatracker.ietf.org/doc/html/rfc7636) used for endorsed platforms (Android, iOS, and web). While every step needed to authorize the app for "authorization code grant with PKCE" happens on a single device, "device flow" requires the user to use a secondary device (PC, tablet, or desktop) to finalize the authorization.

> To access Google Sign-In for Tizen, you'll need to make sure to [register your application for TVs and limited input devices](https://developers.google.com/identity/gsi/web/guides/devices). You also need to add some additional code to your app to fully integrate `google_sign_in` on Tizen (explained below).

Reviewer: The wording is not clear. What does "To access Google Sign-In for Tizen" mean?

Also which information do you want the user to read from the linked page? What should the user do to "register their application"?

Author: How about:

To use `google_sign_in` and sign-in to your account, you first need create credentials for your app ("Client ID" and "Client Secret"). See [register your application for TVs and limited input devices](https://developers.google.com/identity/gsi/web/guides/devices#get_a_client_id_and_client_secret) on how to create them.

You also need to add some additional code to your app to fully integrate `google_sign_in` on Tizen (explained below).

Reviewer: sign-in to your account → This is wrong. The "developer" creates credentials for their app and the "user" of the app uses their account to sign in.

Author: Ah, you're right.

To allow users to sign-in to their account using `google_sign_in`,

### Adding OAuth credentials

> Unlike "authorization code grant with PKCE", "device flow" requires [client secret](https://developers.google.com/identity/protocols/oauth2/limited-input-device#step-4:-poll-googles-authorization-server) parameter during token request. You must call `setCredentials` function with your client's OAuth credentials before calling `google_sign_in`'s API.

Reviewer: - Shouldn't "authorization code grant with PKCE" and "device flow" be capitalized?
- "a" client secret parameter
- You're assuming the reader already knows what "during token request" means. Could you reword if it's not the case?

Author: > Shouldn't "authorization code grant with PKCE" and "device flow" be capitalized?

Yes they should.

> You're assuming the reader already knows what "during token request" means. Could you reword if it's not the case?

How about:

Unlike "Authorization Code Grant with PKCE", "Device Flow" requires a [client secret](https://developers.google.com/identity/protocols/oauth2/limited-input-device#step-4:-poll-googles-authorization-server) parameter during Google Sign-In.

Reviewer: Looks good. Be sure to be consistent throughout the document.

```dart

> import 'package:google_sign_in_tizen/google_sign_in_tizen.dart as tizen'

Reviewer: ```suggestion
import 'package:google_sign_in_tizen/google_sign_in_tizen.dart' as tizen
```

tizen.setCredentials(
  clientId: 'YOUR_CLIENT_ID',
  clientSecret: 'YOUR_CLIENT_SECRET',
);
```

:warning: Security concerns

> Storing client secret in code is considered bad practice as it exposes [security vulnerabilites](https://datatracker.ietf.org/doc/html/rfc8628#section-5.6). Therefore Google allows a [limited set of scopes](https://developers.google.com/identity/protocols/oauth2/limited-input-device#allowedscopes) that can be requested for device flow. Nevertheless, you can perform extra steps to protect your credentials better.

Reviewer: You might refine this paragraph ("Therefore ~ better.")

Author: How about:

Although Google limits API scopes for Device Flow to protect user's sensitive data, you can perform extra steps to protect your client credentials.

1. Do not upload credentials to public repositories.

> Anyone who has access to your credentials can impersonate your app. Make sure to not upload your production credentials in any source code repositories. See the [example app](/example/) for a simple demonstration.

Reviewer: Indent properly by adding 3 spaces.

Also in what aspect the example app demonstrates your proposal?

Author: The comments in `main.dart` line 15~18. If the comments sufficiently explains how real credentials are not being uploaded to the repository, I'll add the file name and lines in this paragraph.

```bash
flutter-tizen build tpk --obfuscate --split-debug-info=/<project-name>/<directory>
```

See [Obfuscating Dart code](https://docs.flutter.dev/deployment/obfuscate) for more information.

### Assigning the plugin's navigatorKey object to MaterialApp or CupertinoApp

During Google Sign-In with device flow, the client displays a widget that shows [user_code](https://developers.google.com/identity/gsi/web/guides/devices#obtain_a_user_code_and_verification_url) and [verification url](https://developers.google.com/identity/gsi/web/guides/devices#obtain_a_user_code_and_verification_url) to the user and instructs them to visit the verification url in a user agent on a secondary device (for example, in a browser on their mobile phone) and enter the user code. The plugin will be able to show this widget only when the plugin's navigatorKey object is assigned to the `navigatorKey` parameter of `MaterialApp` or `CupertinoApp` using the `getNavigatorKey` function.

```dart
import 'package:google_sign_in_tizen/google_sign_in_tizen.dart as tizen'

MaterialApp(
  title: 'Google Sign In',
  navigatorKey: tizen.getNavigatorKey(),
  home: const SignInDemo(),
);
```

> If you need to assign a custom navigatorKey to `MaterialApp`'s (or `CupertinoApp`'s) `navigatorKey` parameter, you must set that key to be used by the plugin as well using `setNavigatorKey` function.

Reviewer: to MaterialApp's navigatorKey parameter → to MaterialApp?

```dart
import 'package:google_sign_in_tizen/google_sign_in_tizen.dart as tizen'

final GlobalKey<NavigatorState> customNavigatorKey = GlobalKey<NavigatorState>(); 

> tizen.setNavigatorKey(customNavigatorKey);

Reviewer: Why don't you use a getter and setter?

```suggestion
tizen.navigatorKey = customNavigatorKey;
```

MaterialApp(
  title: 'Google Sign In',

>   navigatorKey: tizen.getNavigatorKey(),

Reviewer: ```suggestion
  navigatorKey: customNavigatorKey,
```
